proposals/2024-05-23_controller_engine_v2.md

```diff
@@ -0,0 +1,256 @@
+    to support modules of any kind. This is important for code organization and
+    reusability. In order to increase mapping code quality and maintainability,
+    it is important to be able to split the mapping code into multiple files and
+    to be able to reuse code between different mappings. This is especially
+    important now that mapping authors are starting to reuse functionality
+    across different hardware devices (see Numark NS6II and Mixtrack variants or
+    similarity between different pioneer controllers)
+ 3. Current mappings modify global state and assume that they are the only
+    mapping running. This needs to be fixed in order to support multiple
+    mappings running at the same time in the same "execution engine".
+ 4. Device hotplug is currently not possible / hard to implement. This is
+    primarily because of the underlying protocol IO layer not willing to support
+    it. This document proposes an alternative approach that lets us more easily
+    switch the underlying library.
+ 5. Allow changing the manifest format. Many people have expressed distaste in
+    the XML manifest format. Careful implementation of the new format should
+    allow experimentation with different formats.
+ 6. Changing out infrastructure like this is virtually impossible. The proposed
+    architecture is designed to be modular and allows for all the required
+    components to be implemented gradually.
+
+## Goals
+
+Goals and use cases for the solution as proposed in [How](#how):
+
+* Allow multiple controllers to be used at the same time.
+* Mappings should be able to organize themselves into modules.
+* IO protocols should easily be changeable and not not be tied to the mapping.
+* Allow easy, gradual implementation of the system.
+
+### Audience
+
+This is primarily targeted at developers and power users who are interested in
+the controller engine. End-users only interested in the point-and-click mapping
+editor should not experience any significant changes.
+
+## Non-Goals
+
+* Replace `ControllerScriptEngine` in the short term until we are confident in
+  the new design.
+* make changes to the scratching code as it is orthogonal to this proposal.
+* Make changes to the effects system as it is orthogonal to this proposal.
+
+## How
+
+Explain the full overview of the proposed solution. Some guidelines:
+
+### File Structure
+
+Built-in mappings reside in `res/controllers/` where each mapping is a directory
+containing at least a manifest file named `manifest.xxx` (where `.xxx` is the
+language-specific file extension (eg. `.xml`, `.yml`, `.json`, etc)).
+Additionally, `res/controllers/lib` contains shared modules that can be used by
+multiple mappings (such as componentsJS).
+
+### Manifest
+
+The manifest is a declarative file that describes the mapping. It contains the
+following information:
+
+* Sources and Sinks
+* Execution Engines
+  * Used Mixxx APIs
+* Capabilities
```

Comment:
I am afraid this flexibility will introduce lot of boilerplate for no reason. I would prefer to straight forward code a new engine without adding facilities for a not yet known future engine. YAGNI
That does not mean to not allow such future engine, it is always a balance.

Reply:
Do you mean boilerplate in the C++ domain or just in the manifest? For the manifest I could imagine adding some shortcuts for the common usecase sure, but I'd like to keep the decoupled nature on the C++ side. Otherwise we'll be right back where we started with the current engine.